server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1183,145 +1197,155 @@ private void writeUpdatedShardMetaDataAndComputeDeletes(
-                                    ),
-                                    ex
-                                );
-                                // Just passing null here to count down the listener instead of failing it, the stale data left behind
-                                // here will be retried in the next delete or repository cleanup
-                                allShardsListener.onResponse(null);
-                            }
-                        });
+                    for (int shardId = 0; shardId < shardCount; shardId++) {
+                        snapshotExecutor.execute(new ShardSnapshotsDeletion(shardId, survivingSnapshots, allShardsListener));
                     }
                 }));
             }
-        }
 
-        // -----------------------------------------------------------------------------------------------------------------------------
-        // Updating each shard
+            // -----------------------------------------------------------------------------------------------------------------------------
+            // Updating each shard
 
-        /**
-         * Delete snapshot from shard level metadata.
-         *
-         * @param indexGeneration generation to write the new shard level level metadata to. If negative a uuid id shard generation should
-         *                        be used
-         */
-        private ShardSnapshotMetaDeleteResult deleteFromShardSnapshotMeta(
-            Set<SnapshotId> survivingSnapshots,
-            IndexId indexId,
-            int shardId,
-            Collection<SnapshotId> snapshotIds,
-            BlobContainer shardContainer,
-            Set<String> originalShardBlobs,
-            BlobStoreIndexShardSnapshots snapshots,
-            long indexGeneration
-        ) {
-            // Build a list of snapshots that should be preserved
-            final BlobStoreIndexShardSnapshots updatedSnapshots = snapshots.withRetainedSnapshots(survivingSnapshots);
-            ShardGeneration writtenGeneration = null;
-            try {
-                if (updatedSnapshots.snapshots().isEmpty()) {
-                    return new ShardSnapshotMetaDeleteResult(indexId, shardId, ShardGenerations.DELETED_SHARD_GEN, originalShardBlobs);
-                } else {
-                    if (indexGeneration < 0L) {
-                        writtenGeneration = ShardGeneration.newGeneration();
-                        INDEX_SHARD_SNAPSHOTS_FORMAT.write(updatedSnapshots, shardContainer, writtenGeneration.toBlobNamePart(), compress);
+            private class ShardSnapshotsDeletion extends AbstractRunnable {
+
+                private final int shardId;
+                private final Set<SnapshotId> survivingSnapshots;
+                private final ActionListener<ShardSnapshotMetaDeleteResult> allShardsListener;
+                private BlobContainer shardContainer;
+                private Set<String> originalShardBlobs;
```

Comment:
I'm going to leave them as fields for now (I think this will be useful in subsequent reworking of this area) but I added a comment about them to help remind us to rethink this later.